### ISSUE TITLE
how to achieve a long live connections using spark api

### ISSUE THREAD

**Issue by jiacai2050** [elastic/elasticsearch-hadoop#619]
I can write my rdds to es with following java code
```
JavaEsSpark.saveToEs(rdd, indexName);
```
Actually I am using Spark Streaming to write data to es constantly.
In my box, I can see lots of `TIME_WAITED` when execute `netstat -an | grep 9200`.
Clearly, every time the `saveToEs` method gets invoked, it creates a new connection to add data to es.

So is there any way to achieve a long live connections ?

FYI, I found [this link](https://groups.google.com/forum/#!topic/elasticsearch/jbguN90ubJg), but still have no idea how to do.

**Comment by costin** [comment]
`saveToEs` starts a Spark task that in turn, looks at the ES index and then writes the data in parallel.
The `long-lived` connection doesn't really work in this case since this is not just one program where state is passed around but actually a number of jobs that are being fired very rapidly. The `TIME_WAITED` status actually indicates that the connection has closed but it is now waiting on any stray packets that might have not arrived.
This is typically an OS setting - for a more aggressive setup, you can minimize its value.

Back to the issue, there's no much the connector can do at this point - it does not have access to the runtime and it is not possible to pass a connection from Spark to ES-Hadoop (where would that connection be by the way)?

Try to minimize the frequency of `saveToEs` calls by using a bigger/larger RDD. If that is not working for you, than maybe using a different client (maybe the java one) would work better for you instead of es-hadoop.
Basically in that case, you would run everything from within the driver and the whole partitioning of the RDD would not be useful. And everything would run within the same machine since otherwise how would you share your 'connection'...

**Comment by jiacai2050** [comment]
Thanks costin, you said
> where would that connection be by the way)?

I think the connection can be passed in this way
```
JavaEsSpark.saveToEs(conn, rdd, indexName);
```

I have already known es-hadoop use `httpclient` API underneath, so it seems I should meet my demands from scratch as you said above.

**Comment by costin** [comment]
Except it won't work :)
The connection is between points A and B. The saveToES or any RDD operation for that matter triggers a job and thus is distributed. Thus nodes X, Y, Z cannot make use of a connection between A and B (not to mention that connections, HTTP or not, are not stateless and thus cannot/should not be shared).

**Comment by jiacai2050** [comment]
I am not sure how RDD operations are implemented, since they share memory I think they are just  multithread programs. To be exactly, 
> An executor on a host spawns a new thread for a RDD operation.

Speaking of this, I have realized my solution just won't work. If my understanding of RDD operations is correct, then is there any way to init a connection when an executor start up ?
Maybe I should read documentation of spark more carefully instead of asking for help here.
Anyway, many thanks for your help.

**Comment by Junaidsakib** [comment]
Two cents: https://www.elastic.co/guide/en/elasticsearch/hadoop/master/spark.html#spark-streaming

**Comment by jbaiera** [comment]
This is already resolved in 5.0 with our native spark streaming support
